### PR TITLE
feat(installer): preflight probes every download host, cross-OS + hard (#416)

### PR DESCRIPTION
**Pull request by shujaatTracebloc**
Closes #416.

## Problem
Preflight proved connectivity to registries + tracebloc endpoints, then Step 1 downloaded from hosts it *never probed* (`desktop.docker.com`, `dl.k8s.io`, `get.helm.sh`, `github.com`/`objects.githubusercontent.com`, `auth.docker.io`). On TLS-intercepting / allowlist networks the install passed an all-green preflight, then died ~30s later — misleading the IT contact who just watched it pass. Lukas's cross-OS audit widened this from "a few Windows hosts" to a cross-OS preflight gap.

## Fix (both halves, in lockstep)
Probe **every download host the default path fetches from**, per-OS and **HARD** (a blocked one is now a named red preflight line, not a warn) — but only when the fetch will actually happen (tool/app absent; a present tool is never re-downloaded, so its host isn't probed).

| Scope | Hosts added |
|-------|-------------|
| always | `auth.docker.io` (Docker Hub token host) |
| Linux | `get.docker.com`, `download.docker.com`, `github.com` + `objects.githubusercontent.com`, `dl.k8s.io`, `get.helm.sh` |
| macOS | `raw.githubusercontent.com` (Homebrew), `desktop.docker.com` |
| Windows | `desktop.docker.com`, `dl.k8s.io`, `get.helm.sh`, `github.com` + `objects.githubusercontent.com` |

- `objects.githubusercontent.com` is probed **explicitly**: release assets 302 there and `_pf_probe_url` can't follow redirects, so `github.com` passing proved nothing about the asset host.
- **GPU hosts deliberately excluded** — GPU setup is optional and #415 handles its failure with runnable remedies; hard-failing preflight would contradict that.
- New `check-drift.sh` parity check (`_drift_preflight_hosts`) locks the shared-core host set so `preflight.sh` and `install-k8s.ps1` can't drift apart.

## Acceptance criteria
- [x] Any blocked download host = a named red preflight line **before anything is installed or modified**
- [x] bash + ps1 preflight host lists **covered by the drift check**

## Tests
+5 bats (`preflight.bats`), +2 bats (`check-drift.bats`), +7 Pester (`install-k8s.Tests.ps1`).

**One existing test updated:** the single warn-only assertion in `preflight.bats` is updated to the new hard behaviour (the acceptance criterion requires red/hard). It's the only existing test touched — everything else is additive.

Verified locally (CI-identical):
- Pester full: **182 passed, 0 failed**, 8 skipped (Windows-only)
- `preflight.bats` + `check-drift.bats`: **74 passed, 0 failed**
- all bats: 3 failures are pre-existing macOS-local flakes (confirmed on clean `develop`); ubuntu CI is the arbiter
- PSScriptAnalyzer: **0 Error-severity** · check-style / drift: clean · manifest regenerated

---

> [!NOTE]
> **Low Risk**
> Changes are limited to installer preflight/drift checks and tests; no runtime cluster, auth, or data-path behavior.
> 
> **Overview**
> Fixes the gap where preflight could pass on corporate/allowlist networks, then the install failed ~30s later on hosts that were never probed (#416).
> 
> **Connectivity:** Adds **`auth.docker.io`** as a hard critical probe on Windows (`install-k8s.ps1`) and Unix (`preflight.sh`). Promotes tool-download hosts from warn-only to **hard failures** when that tool is missing—e.g. `dl.k8s.io`, `get.helm.sh`, `github.com`, `objects.githubusercontent.com`, and on Windows `desktop.docker.com` when Docker Desktop isn’t installed. Probes run only when a fetch will happen (installed tools aren’t re-checked). Linux keeps Docker-engine install URLs (`get.docker.com` / `download.docker.com`) **soft** (warn) because the path varies by distro; macOS branches on **`_pf_has_gui_session`** (Docker Desktop vs Colima/brew) so blocked hosts on unused paths don’t abort installs.
> 
> **Parity & tests:** New **`check-drift.sh`** Check 4 (`_drift_preflight_hosts`) compares probe URLs between `preflight.sh` and `install-k8s.ps1` (not comment/hint greps). **`manifest.sha256`** updated for changed scripts. Expanded **bats** (`preflight.bats`, `check-drift.bats`) and **Pester** (`install-k8s.Tests.ps1`) cover hard fails, macOS GUI/headless branches, and drift edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a71710434b7be4cd55cbb33142f516e1bc09cd09. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: scripts/lib/preflight.sh [hunk 1]
```diff
@@ -392,12 +392,34 @@ _pf_connectivity() {
...
+  if [[ "$OS" == "Linux" ]]; then
+    if ! has docker;  then criticals+=("Docker install (get.docker.com)|https://get.docker.com/" \
+                                       "Docker packages (download.docker.com)|https://download.docker.com/"); fi
+    if ! has k3d;     then criticals+=("k3d download (github.com)|https://github.com/" \
+                                       "k3d assets (objects.githubusercontent.com)|https://objects.githubusercontent.com/"); fi
+    if ! has kubectl; then criticals+=("kubectl (dl.k8s.io)|https://dl.k8s.io/"); fi
+    if ! has helm;    then criticals+=("Helm (get.helm.sh)|https://get.helm.sh/"); fi
+  elif [[ "$OS" == "Darwin" ]]; then
+    if ! has brew;    then criticals+=("Homebrew install (raw.githubusercontent.com)|https://raw.githubusercontent.com/"); fi
+    if ! has docker;  then criticals+=("Docker Desktop (desktop.docker.com)|https://desktop.docker.com/"); fi
```
[the remark this answers is not in the document]
Comment (shujaatTracebloc) [reply scripts/lib/preflight.sh]: Fixed in fcdaf2b. `desktop.docker.com` is now **warn-only** on macOS: headless Macs use Colima via brew (`ghcr.io`, already hard-probed), so a blocked Desktop CDN no longer aborts that supported path. Only the k8s tool binaries + registries stay hard. (Windows keeps it hard — Docker Desktop is the sole path there.)

File: scripts/lib/preflight.sh [hunk 2]
```diff
@@ -434,26 +456,6 @@ _pf_connectivity() {
...
-      label="${c%%|*}"; url="${c#*|}"
-      status="$(_pf_probe_url "$url")"
-      if [[ "$status" == "ok" ]]; then
-        _pf_ok "${label} reachable"
-      else
-        warn "${label} unreachable (${status}) — needed only to install that tool."
-      fi
-    done
-  fi
-
```
[the remark this answers is not in the document]
Comment (shujaatTracebloc) [reply scripts/lib/preflight.sh:456]: Fixed in fcdaf2b. The egress hint now names the tool-download hosts (`dl.k8s.io` / `get.helm.sh` / `github.com` / `objects.githubusercontent.com`) alongside the always-critical set — in both `preflight.sh` and `install-k8s.ps1` — so a red line for a blocked download host has matching remediation.

File: scripts/lib/preflight.sh [hunk 3]
```diff
@@ -392,12 +392,34 @@ _pf_connectivity() {
...
+  # Download hosts the default install path fetches from on THIS OS — promoted to
+  # HARD (#416): a blocked one used to pass preflight then fail the install ~30s
+  # later. Only added when the fetch will actually happen (tool/app absent; a
+  # present tool is never re-downloaded, so its host is not probed). Release assets
+  # 302 to objects.githubusercontent.com and _pf_probe_url does not follow
+  # redirects, so github.com passing proves nothing about the asset host — probe it
+  # explicitly. Kept in lockstep with install-k8s.ps1 (drift check: check-drift.sh).
+  if [[ "$OS" == "Linux" ]]; then
+    if ! has docker;  then criticals+=("Docker install (get.docker.com)|https://get.docker.com/" \
+                                       "Docker packages (download.docker.com)|https://download.docker.com/"); fi
```
[the remark this answers is not in the document]
Comment (shujaatTracebloc) [reply scripts/lib/preflight.sh]: Fixed in fcdaf2b. `get.docker.com` / `download.docker.com` are now **warn-only** (Docker-engine host is path/distro-dependent — pacman/zypper/Amazon use distro repos, RHEL clones use only download.docker.com). A blocked host on a path that won't use it no longer aborts the install; the unambiguous k8s tool hosts stay hard.

File: scripts/lib/preflight.sh [hunk 4]
```diff
@@ -392,12 +392,43 @@ _pf_connectivity() {
...
+    if ! has k3d;     then criticals+=("k3d download (github.com)|https://github.com/" \
+                                       "k3d assets (objects.githubusercontent.com)|https://objects.githubusercontent.com/"); fi
+    if ! has kubectl; then criticals+=("kubectl (dl.k8s.io)|https://dl.k8s.io/"); fi
+    if ! has helm;    then criticals+=("Helm (get.helm.sh)|https://get.helm.sh/"); fi
+    if ! has docker;  then soft+=("Docker install (get.docker.com)|https://get.docker.com/" \
+                                  "Docker packages (download.docker.com)|https://download.docker.com/"); fi
+  elif [[ "$OS" == "Darwin" ]]; then
+    if ! has brew;    then criticals+=("Homebrew install (raw.githubusercontent.com)|https://raw.githubusercontent.com/"); fi
+    if ! has docker;  then soft+=("Docker Desktop (desktop.docker.com)|https://desktop.docker.com/"); fi
+    # k3d / kubectl / helm install via Homebrew bottles on ghcr.io (probed above).
```
[the remark this answers is not in the document]
Comment (shujaatTracebloc) [reply scripts/lib/preflight.sh]: Fixed in ddcf239. On macOS when brew is absent, preflight now hard-probes github.com (the Homebrew/brew + core clone host) in addition to raw.githubusercontent.com (the install-script host) — closing the same allow-raw/block-github gap. Added a preflight.bats case.

**Review by saadqbal — COMMENTED** [review]
Careful PR — the conditional/hard-vs-soft logic is right, and I traced the Linux + Windows fetch hosts against setup-linux.sh / install-k8s.ps1 and every hard-probed host matches what the installer actually pulls from. Two things worth a look before merge (inline): the drift check doesn't actually catch a removed probe, and macOS misses formulae.brew.sh. Neither blocks the core fix.

File: scripts/tests/check-drift.sh [hunk 5]
```diff
@@ -186,12 +186,40 @@ _drift_cli_contract() {
...
+  local shared=(
+    registry-1.docker.io auth.docker.io ghcr.io
+    dl.k8s.io get.helm.sh github.com objects.githubusercontent.com
+    desktop.docker.com tracebloc.github.io
+  )
+  local before=$_drift f h
+  for f in "${files[@]}"; do
+    if [[ ! -f "$DRIFT_ROOT/$f" ]]; then _note "$f is missing"; continue; fi
+    for h in "${shared[@]}"; do
+      grep -qF -- "$h" "$DRIFT_ROOT/$f" || \
```
Comment (saadqbal) [inline scripts/tests/check-drift.sh]: This greps the whole file, so it also matches these hosts inside comments and the egress-hint strings — and every shared host appears in both preflight.sh and install-k8s.ps1's hints/comments. So the check passes even if you delete an actual probe line; it can't detect the drift it's meant to (the AC "host lists covered by the drift check" isn't really enforced). Better to extract just the probed URLs (the `criticals`/`$criticals` entries) and diff those. The two new bats cases only go green because the fixtures don't carry the hint/comment text the real files do.

File: scripts/lib/preflight.sh [hunk 6]
```diff
@@ -392,12 +392,46 @@ _pf_connectivity() {
...
+    if ! has helm;    then criticals+=("Helm (get.helm.sh)|https://get.helm.sh/"); fi
+    if ! has docker;  then soft+=("Docker install (get.docker.com)|https://get.docker.com/" \
+                                  "Docker packages (download.docker.com)|https://download.docker.com/"); fi
+  elif [[ "$OS" == "Darwin" ]]; then
+    # Homebrew install fetches the script from raw.githubusercontent.com AND then
+    # git-clones Homebrew/brew + core from github.com — probe both (Bugbot #416).
+    if ! has brew;    then criticals+=("Homebrew install (raw.githubusercontent.com)|https://raw.githubusercontent.com/" \
+                                       "Homebrew clone (github.com)|https://github.com/"); fi
+    if ! has docker;  then soft+=("Docker Desktop (desktop.docker.com)|https://desktop.docker.com/"); fi
+    # k3d / kubectl / helm install via Homebrew bottles on ghcr.io (probed above).
```
Comment (saadqbal) [inline scripts/lib/preflight.sh]: `brew install kubectl/helm/k3d` (and colima/docker) pulls formula metadata from formulae.brew.sh, not only bottles from ghcr.io — and that host isn't probed. When brew is already present (common), ghcr.io is the only tool host checked, so a blocked formulae.brew.sh = green preflight then a failed `brew install`: the exact #416 gap, on macOS.

File: scripts/tests/check-drift.sh (hunk 5 above)
Comment (shujaatTracebloc) [reply scripts/tests/check-drift.sh]: Good catch — fixed in c28d5a6. The check now extracts only the hosts in an actual probe URL (`_drift_probed_hosts_sh` greps `|https://host` in preflight.sh; `_drift_probed_hosts_ps1` greps `url = "https://host` in install-k8s.ps1) and diffs those, instead of grepping the whole file — so a deleted probe line is now detected even if the host lingers in a comment or the egress hint. `tracebloc.github.io` drops from the shared set (ps1 probes it via $TRACEBLOC_HELM_REPO_URL, not a literal; Check 1 already pins the host map). Rewrote the two bats fixtures to real probe entries and added a case that keeps a host in a comment+hint but removes its probe URL — it now correctly flags drift (would have passed before).

File: scripts/lib/preflight.sh (hunk 6 above)
Comment (shujaatTracebloc) [reply scripts/lib/preflight.sh]: Fixed in c28d5a6. macOS now hard-probes `formulae.brew.sh` whenever a brew-installed tool (kubectl/k3d/helm/docker) is absent — independent of whether brew itself is present — so the metadata host is covered, not just the ghcr.io bottles. Added a preflight.bats case for the brew-present/tools-absent scenario you described.

File: scripts/tests/check-drift.sh [hunk 7]
```diff
@@ -186,12 +186,50 @@ _drift_cli_contract() {
...
+# comments and the egress-hint strings, so it would pass even if a real probe line
+# were deleted (reviewer: it couldn't detect the drift it exists to catch).
+#
+# The shared-core set is every host BOTH probe as a URL literal on the default
+# path. OS-specific hosts (get.docker.com / download.docker.com on Linux,
+# raw.githubusercontent.com / formulae.brew.sh on macOS) are excluded, as is
+# tracebloc.github.io (install-k8s.ps1 probes it via the $TRACEBLOC_HELM_REPO_URL
+# variable, not a literal — Check 1 already pins the backend/chart host map).
+_drift_probed_hosts_sh()  { grep -oE '\|https://[a-zA-Z0-9.-]+' "$1" 2>/dev/null | sed 's#.*//##' | sort -u; }
+_drift_probed_hosts_ps1() { grep -oE 'url *= *"https://[a-zA-Z0-9.-]+' "$1" 2>/dev/null | sed 's#.*//##' | sort -u; }
```
[the remark this answers is not in the document]
Comment (shujaatTracebloc) [reply scripts/tests/check-drift.sh]: Fixed in 7384b12. The ps1 host extractor now requires `label =` on the same line (the shape of a preflight $criticals entry), so an unrelated `$url = "https://github.com/…"` download line (the winget bootstrap) no longer counts — deleting the real k3d github.com probe now flags drift. Added a check-drift.bats case with a stray $url= line + a removed probe that correctly fails.

File: scripts/lib/preflight.sh [hunk 8]
```diff
@@ -392,12 +392,53 @@ _pf_connectivity() {
...
+    if ! has brew; then criticals+=("Homebrew install (raw.githubusercontent.com)|https://raw.githubusercontent.com/" \
+                                    "Homebrew clone (github.com)|https://github.com/"); fi
+    # `brew install` (kubectl/k3d/helm, and colima/docker) pulls formula METADATA
+    # from formulae.brew.sh — bottles come from ghcr.io (probed above), but the
+    # metadata host is separate and is hit even when brew is already installed. A
+    # blocked formulae.brew.sh = green preflight then a failed `brew install`
+    # (reviewer, #416). Probe it whenever a brew-installed tool is missing.
+    if ! has kubectl || ! has k3d || ! has helm || ! has docker; then
+      criticals+=("Homebrew formulae (formulae.brew.sh)|https://formulae.brew.sh/")
+    fi
```
[the remark this answers is not in the document]
Comment (shujaatTracebloc) [reply scripts/lib/preflight.sh:459]: Fixed in 7384b12. formulae.brew.sh is now probed only on `! has kubectl || ! has k3d || ! has helm` (the tools that always install via brew) — dropped `! has docker`, since a GUI Mac installs Docker Desktop from desktop.docker.com, not brew, so docker-only-missing no longer hard-fails the metadata host. Added a preflight.bats case for docker-only-missing (tools present) asserting formulae.brew.sh is not probed.

File: scripts/lib/preflight.sh [hunk 9]
```diff
@@ -392,12 +392,55 @@ _pf_connectivity() {
...
+    # — bottles come from ghcr.io (probed above), but the metadata host is separate
+    # and is hit even when brew is already installed. A blocked formulae.brew.sh =
+    # green preflight then a failed `brew install` (reviewer, #416). Trigger only on
+    # the tools that ALWAYS install via brew — NOT docker: on a GUI Mac docker comes
+    # from Docker Desktop (desktop.docker.com), so `! has docker` alone would
+    # hard-fail formulae.brew.sh on an install that never runs brew (Bugbot).
+    if ! has kubectl || ! has k3d || ! has helm; then
+      criticals+=("Homebrew formulae (formulae.brew.sh)|https://formulae.brew.sh/")
+    fi
+    if ! has docker; then soft+=("Docker Desktop (desktop.docker.com)|https://desktop.docker.com/"); fi
```
[the remark this answers is not in the document]
Comment (shujaatTracebloc) [reply scripts/lib/preflight.sh]: Fixed in 8912268. formulae.brew.sh is now probed for a missing docker only when there is no GUI session (_pf_has_gui_session, mirroring setup-macos.sh): headless Macs install colima/docker via brew and hit formulae.brew.sh (now hard-probed); GUI Macs use Docker Desktop and still skip it. Added bats cases for both GUI and headless paths.

File: scripts/lib/preflight.sh [hunk 10]
```diff
@@ -392,12 +402,56 @@ _pf_connectivity() {
...
+    # when brew is already installed. A blocked formulae.brew.sh = green preflight
+    # then a failed `brew install` (reviewer, #416). kubectl/k3d/helm ALWAYS install
+    # via brew. docker is path-dependent: a GUI Mac installs Docker Desktop from
+    # desktop.docker.com (no brew — probing formulae would false-fail, Bugbot r3),
+    # but a HEADLESS Mac installs colima/docker via brew (formulae IS needed,
+    # Bugbot r4). So add docker to the trigger only when there's no GUI session.
+    if ! has kubectl || ! has k3d || ! has helm || { ! has docker && ! _pf_has_gui_session; }; then
+      criticals+=("Homebrew formulae (formulae.brew.sh)|https://formulae.brew.sh/")
+    fi
+    if ! has docker; then soft+=("Docker Desktop (desktop.docker.com)|https://desktop.docker.com/"); fi
```
[the remark this answers is not in the document]
Comment (shujaatTracebloc) [reply scripts/lib/preflight.sh]: Fixed in a717104. Consolidated the macOS Docker-path probing into one GUI-aware block: docker missing + GUI session now hard-probes desktop.docker.com (the actual Docker Desktop path) instead of warn-only; docker missing + headless routes to formulae.brew.sh (colima/docker via brew) and does not probe desktop.docker.com. desktop.docker.com is no longer warn-only. Added GUI/headless bats cases for both hosts.

**Review by saadqbal — APPROVED** [review]
